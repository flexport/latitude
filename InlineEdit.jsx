/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import invariant from "./tools/invariant";
import {ENTER} from "./constants/interactions/KeyCodes";
import {
  HORIZONTAL_INPUT_PADDING_SMALL_PX,
  HORIZONTAL_INPUT_PADDING_BASE_PX,
  HORIZONTAL_INPUT_PADDING_LARGE_PX,
} from "./styles/input";
import IconButton from "./button/IconButton";
import Button from "./button/Button";
import Text from "./Text";
import useSetTimeout from "./hooks/useSetTimeout";
import {commonT as t} from "./config/I18n";

export type RenderProp<T> = (props: {
  +value: T,
  +placeholder?: string,
  +onChange: (newValue: T) => void,
  +inputRef: (HTMLElement | null) => void,
  +onKeyDown: (e: KeyboardEvent) => void,
}) => React.Node;

type Props<T> = {|
  /** The value to be displayed when in display mode */
  +value: T,
  /**
   * A placeholder value to display if value is unset.
   * Text inputs cannot be unset unless a placeholder is given
   */
  +placeholder?: string,
  /** Called when the user has approved a change to the input's value */
  +onChange: (newValue: T) => void,
  /** Determines the style of edit button to be displayed */
  +editStyle?: "button" | "pencil",
  /** Whether the display mode text will wrap or overflow */
  +textWrap?: boolean,
  +size?: "s" | "m" | "l",
  /** A renderProp that will render the input with the appropriate props */
  +children: RenderProp<T>,
  /** A static React Node that will be displayed when in display Mode */
  +staticView?: React.Node,
|};

const TRANSITION_DURATION = 150;

/**
 * @short A wrapper component that transforms an input into an inline-editable input.
 * @category Data Entry
 * @brandStatus V2
 * @status Beta
 */
function InlineEdit<T>({
  value,
  placeholder,
  onChange,
  editStyle = "button",
  textWrap = false,
  size = "m",
  children,
  staticView,
}: Props<T>) {
  const lastInputRef = React.useRef(null);

  const [editMode, setEditMode] = React.useState(false);
  const [editText, setEditText] = React.useState(value);

  // used to keep track of transition states between display mode and edit mode
  const [transitioningIn, setTransitioningIn] = React.useState(false);
  const [transitioningOut, setTransitioningOut] = React.useState(false);

  const setTimeout = useSetTimeout();

  const handleEditClick = React.useCallback(() => {
    setEditText(value);
    setTransitioningIn(true);

    setTimeout(() => {
      setEditMode(true);
      lastInputRef.current = null;
      setTransitioningIn(false);
    }, TRANSITION_DURATION);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);

  const handleCancelClick = React.useCallback(() => {
    setEditText(value);
    setTransitioningOut(true);

    setTimeout(() => {
      setEditMode(false);
      setTransitioningOut(false);
    }, TRANSITION_DURATION);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);

  const handleCheckClick = React.useCallback(() => {
    setTransitioningOut(true);

    setTimeout(() => {
      setEditMode(false);
      setTransitioningOut(false);
    }, TRANSITION_DURATION);

    if (
      typeof editText === "string" &&
      editText.trim() === "" &&
      !placeholder
    ) {
      return;
    }

    onChange(editText);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [editText]);

  let content = null;

  const stringifiedValue = String(value);
  const valueOrPlaceholder = stringifiedValue || placeholder;
  const shouldShowPlaceholder = !stringifiedValue;

  if (editMode && !(transitioningIn || transitioningOut)) {
    content = (
      <div className={css(styles.editableContainer)}>
        {children({
          value: editText,
          onChange: setEditText,
          placeholder,
          onKeyDown: (e: KeyboardEvent) => {
            if (e.keyCode === ENTER && !e.shiftKey) {
              handleCheckClick();
            }
          },
          inputRef: inputRef => {
            if (inputRef !== null) {
              if (!lastInputRef.current) {
                // $FlowFixMe(Dirak) inputRef has select()
                inputRef.select();
              }

              lastInputRef.current = inputRef;
            }
          },
        })}

        <div className={css(styles.actionButtonsContainer)}>
          <div style={{marginRight: "4px"}}>
            <IconButton
              iconName="cancel"
              type="button"
              size="s"
              onClick={handleCancelClick}
            />
          </div>
          <IconButton
            iconName="check"
            type="button"
            intent="basic"
            size="s"
            onClick={handleCheckClick}
          />
        </div>
      </div>
    );
  } else if (editStyle === "button") {
    content = (
      <>
        <div
          className={css(
            styles.editButtonValueContainer,
            textWrap ? styles.wrap : styles.noWrap
          )}
        >
          {staticView || (
            <Text color={shouldShowPlaceholder ? "grey40" : "grey60"}>
              {valueOrPlaceholder}
            </Text>
          )}
        </div>
        <div
          className={css(
            styles.editButtonContainer,
            (transitioningIn || transitioningOut) && styles.hidden
          )}
        >
          <Button intent="basic" kind="blank" onClick={handleEditClick}>
            {t("Edit")}
          </Button>
        </div>
      </>
    );
  } else if (editStyle === "pencil") {
    invariant(
      textWrap,
      "InlineEditable with pencil editStyle must have prop textWrap set to true"
    );

    content = (
      <div className={css(styles.pencilContentContainer)}>
        <div
          className={css(
            styles.penciledValueContainer,
            textWrap ? styles.wrap : styles.noWrap
          )}
        >
          {staticView || (
            <Text
              color={shouldShowPlaceholder ? "grey40" : "grey60"}
              display="inline"
            >
              {valueOrPlaceholder}
            </Text>
          )}
        </div>
        <div
          className={css(
            styles.pencilIconContainer,
            (transitioningIn || transitioningOut) && styles.hidden
          )}
        >
          <IconButton
            iconName="pencil"
            intent="none"
            kind="blank"
            type="button"
            onClick={handleEditClick}
          />
        </div>
      </div>
    );
  } else {
    throw new Error(
      "InlineEditable editStyle must either be 'button' or 'pencil'"
    );
  }

  const horizontalPadding = getHorizontalPadding(size);

  const transitionInStyle = {
    transition: "transform 75ms ease-in-out",
    transform: `translateX(${horizontalPadding}px)`,
  };

  const transitionOutStyle = {
    transition: "transform 75ms ease-in-out",
    transform: `translateX(-${horizontalPadding}px)`,
    marginLeft: horizontalPadding,
  };

  let style;
  if (transitioningOut) {
    style = transitionOutStyle;
  } else if (transitioningIn) {
    style = transitionInStyle;
  } else {
    style = {};
  }

  return (
    <div className={css(styles.container)} style={style}>
      {content}
    </div>
  );
}

const getHorizontalPadding = (size: "s" | "m" | "l"): number => {
  switch (size) {
    case "s": {
      return HORIZONTAL_INPUT_PADDING_SMALL_PX;
    }
    case "m": {
      return HORIZONTAL_INPUT_PADDING_BASE_PX;
    }
    case "l": {
      return HORIZONTAL_INPUT_PADDING_LARGE_PX;
    }
    default: {
      return HORIZONTAL_INPUT_PADDING_BASE_PX;
    }
  }
};

const styles = StyleSheet.create({
  container: {
    boxSizing: "border-box",
    display: "flex",
    flexDirection: "row",
    alignItems: "flex-start",
    position: "relative",
    width: "100%",
    paddingRight: "30px",
    lineHeight: "18px",
    minHeight: "30px",
  },

  hidden: {
    display: "none",
  },

  editableContainer: {
    display: "flex",
    flexDirection: "column",
    position: "relative",
    flex: "1",
  },

  actionButtonsContainer: {
    display: "flex",
    justifyContent: "flex-end",
    marginTop: "4px",
    position: "absolute",
    width: "100%",
    top: "100%",
    zIndex: "1",
  },

  editButtonContainer: {
    position: "absolute",
    top: "0",
    right: "0",
    padding: `6px 0px 4px 0px`,
  },

  wrap: {
    whiteSpace: "pre-line",
  },

  noWrap: {
    textOverflow: "ellipsis",
    overflow: "hidden",
    whiteSpace: "nowrap",
  },

  editButtonValueContainer: {
    minHeight: "30px",
    border: "1px solid transparent",
    padding: `6px 0px 4px 0px`,
  },

  pencilContentContainer: {
    position: "relative",
    width: "100%",
    border: "1px solid transparent",
    padding: `6px 0px 4px 0px`,
  },

  penciledValueContainer: {
    display: "inline",
    position: "relative",
    boxDecorationBreak: "clone",
  },

  pencilIconContainer: {
    position: "absolute",
    display: "inline",
    marginLeft: "8px",
    padding: `2px 0px 4px 0px`,
  },
});

export default InlineEdit;
