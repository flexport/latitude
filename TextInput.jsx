/**
 * TEAM: frontend_infra
 * WATCHERS: uforic
 * @flow strict
 */

import {StyleSheet, css} from "aphrodite";
import * as React from "react";
import InputGroupContext, {
  CENTER_INPUT,
  LEFT_INPUT,
  RIGHT_INPUT,
} from "./context/InputGroupContext";
import latitudeColors from "./colors";
import {LabelContext} from "./Label";
import IconButton from "./button/IconButton";
import Icon, {type IconNames, type StandardIconSizes} from "./Icon";
import {type Size} from "./sizes";
import {
  getInputStyles,
  HORIZONTAL_INPUT_PADDING_BASE_PX,
  HORIZONTAL_INPUT_PADDING_LARGE_PX,
  HORIZONTAL_INPUT_PADDING_SMALL_PX,
} from "./styles/input";
import iconSizes from "./iconSizes";
import {getWidthOfText} from "./styles/fontWidths";
import {typeScale} from "./styles/typography";

export type TextAlignment = "left" | "right" | "center";
export type TextInputType = "text" | "password" | "email";

export type InputPrefixSuffix =
  | ?string
  | {|iconName: IconNames, onClick?: () => void, onFocus?: () => void|};

export type TextInputProps = {|
  /** note that this must be a string, and cannot be null or undefined. */
  +value: string,
  /** the placeholder text that will be displayed when the input is empty */
  +placeholder: string,
  /** whether the input is disabled or not */
  +disabled: boolean,
  /** whether the input is readonly or not */
  +readOnly: boolean,
  /** whether the input is invalid or not */
  +isInvalid: boolean,
  /** whether the input is prefilled or not */
  +isPrefilled: boolean,
  /** whether the input should show ellipsis and a tooltip on overflow  */
  +textOverflow?: "clip" | "ellipsis",
  /** the size of the input */
  +size: Size,
  /** forwards a ref to the inputfield of the input */
  +inputRef?: ((HTMLElement | null) => void) | {|current: ?HTMLElement|},
  /** called when the value of the textinput changes */
  +onChange: string => void,
  /** called on the mouseDown event */
  +onMouseDown?: Event => void,
  /** called when the keydown event is trigged on the input */
  +onClick?: Event => void,
  /** called when the input is focused */
  +onFocus?: Event => void,
  /** called when the input is blurred */
  +onBlur?: Event => void,
  /** called when the input is clicked */
  +onKeyDown?: KeyboardEvent => void,
  /** most Flexport forms do not use the browser FormData API, which requires
   * named inputs; only use this if you need the form data API to work. */
  +name?: string,
  /** sets the alignment of the text in the input field */
  +textAlign: TextAlignment,
  /** the HTML type of this text input, see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input for a complete list */
  +type: TextInputType,
  /** the maximum number of characters the textinput will accept */
  +maxLength?: number,
  /** prefix sigil that will appear within and at the start of the input field */
  +prefix?: InputPrefixSuffix,
  /** suffix sigil that will appear within and at the end of the input field */
  +suffix?: InputPrefixSuffix,
|};

type SigilProps = {|
  +side: "left" | "right",
  +paddingDefaultTextInput: number,
  +sigil: InputPrefixSuffix,
  +size: Size,
  +disabled: boolean,
|};

/**
 * Sigil returns the appropriate prefix/suffix label based on the type of context obj
 */
function Sigil({
  side,
  paddingDefaultTextInput,
  sigil,
  size,
  disabled,
}: SigilProps) {
  let innerContent = null;

  if (sigil == null) {
    return null;
  }

  if (typeof sigil === "string") {
    innerContent = (
      <span
        className={css(
          styles.sigilText,
          disabled ? styles.sigilTextDisabled : undefined
        )}
        style={{
          ...getTextTypeScale(size),
        }}
      >
        {sigil}
      </span>
    );
  } else if (sigil.onClick) {
    innerContent = (
      <IconButton
        {...sigil}
        disabled={disabled}
        kind="blank"
        intent="none"
        type="button"
        size={size}
      />
    );
  } else {
    innerContent = (
      <Icon
        iconName={sigil.iconName}
        size={getIconSizeFromTextInputSize(size)}
        color="grey50"
      />
    );
  }

  return (
    <span
      className={css(styles.sigil)}
      style={{
        [(side: string)]: `${paddingDefaultTextInput}px`,
        ...(typeof sigil === "string" || !sigil.onClick
          ? {
              pointerEvents: "none",
            }
          : {}),
      }}
    >
      {innerContent}
    </span>
  );
}

/**
 * @category Data Entry
 * @short Collect simple text input from the user.
 * @brandStatus V2
 * @status Stable
 * You may want to use TextInput with InputError or TextInputPrefixSuffix, decorators that
 * modify the look and feel of TextInput. If you need a multiline input, think about using
 * TextareaInput.
 */
function TextInput(props: TextInputProps) {
  const labelContext = React.useContext(LabelContext);
  const inputGroupContext = React.useContext(InputGroupContext);

  const handleFocus = (event: Event) => {
    if (labelContext.labelOnFocus) {
      labelContext.labelOnFocus();
    }

    if (props.onFocus) {
      props.onFocus(event);
    }
  };

  const handleBlur = (event: Event) => {
    if (labelContext.labelOnBlur) {
      labelContext.labelOnBlur();
    }

    if (props.onBlur) {
      props.onBlur(event);
    }
  };

  const {
    size,
    disabled,
    isInvalid,
    isPrefilled,
    textOverflow = "clip",
    onKeyDown,
    onChange,
    onMouseDown,
    placeholder,
    textAlign,
    name,
    value,
    maxLength,
    onClick,
    readOnly,
    prefix = null,
    suffix = null,
  } = props;

  const paddingDefaultTextInput = getPaddingFromTextInputSize(size);

  let prefixPadding = paddingDefaultTextInput;
  let prefixSigil = null;

  if (prefix !== null) {
    prefixPadding = paddingDefaultTextInput * 2 + getWidth(prefix, size);

    prefixSigil = (
      <Sigil
        side="left"
        sigil={prefix}
        size={size}
        paddingDefaultTextInput={paddingDefaultTextInput}
        disabled={disabled}
      />
    );
  }

  let suffixPadding = paddingDefaultTextInput;
  let suffixSigil = null;

  if (suffix !== null) {
    suffixPadding = paddingDefaultTextInput * 2 + getWidth(suffix, size);

    suffixSigil = (
      <Sigil
        side="right"
        sigil={suffix}
        size={size}
        paddingDefaultTextInput={paddingDefaultTextInput}
        disabled={disabled}
      />
    );
  }

  const textFieldStyles = {
    textAlign,
    paddingLeft: prefixPadding,
    paddingRight: suffixPadding,
  };

  const inputStyle = css(
    ...getInputStyles({
      size,
      readOnly,
      disabled,
      isInvalid,
      isPrefilled,
      applyEllipsis: textOverflow === "ellipsis",
      noPadding: true,
    }),
    inputGroupContext === CENTER_INPUT && styles.noBorders,
    inputGroupContext === LEFT_INPUT && styles.noRightBorders,
    inputGroupContext === RIGHT_INPUT && styles.noLeftBorders
  );

  const passwordAtts =
    props.type === "password"
      ? {autoComplete: "off", "data-sensitve": true}
      : undefined;

  return (
    <div className={css(styles.wrapper)}>
      <input
        className={inputStyle}
        disabled={disabled}
        onKeyDown={onKeyDown}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onChange={evt => onChange(evt.target.value)}
        onMouseDown={onMouseDown}
        ref={props.inputRef}
        onClick={onClick}
        type={props.type}
        placeholder={placeholder}
        value={value}
        name={name}
        maxLength={maxLength}
        style={textFieldStyles}
        readOnly={readOnly}
        {...passwordAtts}
      />
      {prefixSigil}
      {suffixSigil}
    </div>
  );
}

TextInput.defaultProps = {
  disabled: false,
  placeholder: "",
  size: "m",
  isPrefilled: false,
  isInvalid: false,
  textAlign: "left",
  maxLength: 255,
  readOnly: false,
  type: "text",
};

export default TextInput;

/**
 * getWidth returns the appropriate width for text or icon depending on the text, icon name and size we provide since icons and text varies in size
 */
const getWidth = (sigil: InputPrefixSuffix, size: Size) => {
  switch (typeof sigil) {
    case "string": {
      return getWidthOfText(sigil);
    }
    default: {
      return getSvgSizeFromTextInputSize(size);
    }
  }
};

const getTextTypeScale = (size: Size) => {
  switch (size) {
    case "s":
      return typeScale.subtext;
    case "m":
      return typeScale.base;
    case "l":
      // Need to make custom font size here due to the original font size not existing in constants
      return styles.largeFontSize;
    default:
      return typeScale.base;
  }
};
const getIconSizeFromTextInputSize = (size: Size): StandardIconSizes => {
  // eslint-disable-next-line default-case
  switch (size) {
    case "s":
    case "m":
      return "xs";
    case "l":
      return "s";
  }
  return "xs";
};
const getPaddingFromTextInputSize = (size: Size): number => {
  // eslint-disable-next-line default-case
  switch (size) {
    case "s":
      return HORIZONTAL_INPUT_PADDING_SMALL_PX;
    case "m":
      return HORIZONTAL_INPUT_PADDING_BASE_PX;
    case "l":
      return HORIZONTAL_INPUT_PADDING_LARGE_PX;
  }
  return HORIZONTAL_INPUT_PADDING_BASE_PX;
};

const getSvgSizeFromTextInputSize = (size: Size): number => {
  // eslint-disable-next-line default-case
  switch (size) {
    case "s":
      return iconSizes.xxs;
    case "m":
      return iconSizes.xs;
    case "l":
      return iconSizes.s;
  }
  return iconSizes.xs;
};

const styles = StyleSheet.create({
  largeFontSize: {
    fontSize: "16px",
    lineHeight: "18px",
  },
  wrapper: {
    display: "flex",
    position: "relative",
  },
  sigil: {
    position: "absolute",
    top: "50%",
    transform: "translateY(-50%)",
  },
  sigilTextDisabled: {
    color: latitudeColors.grey40,
  },
  sigilText: {
    lineHeight: "100%",
    color: latitudeColors.grey50,
  },
  noBorders: {
    borderRadius: "0px 0px 0px 0px",
  },
  noLeftBorders: {
    borderRadius: "0px 3px 3px 0px",
  },
  noRightBorders: {
    borderRadius: "3px 0px 0px 3px",
  },
});
