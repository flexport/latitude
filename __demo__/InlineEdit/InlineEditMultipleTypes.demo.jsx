/**
 * TEAM: frontend_infra
 * @flow
 */

import React, {useState} from "react";

import Group from "../../Group";
import InlineEdit from "../../InlineEdit";
import Label from "../../Label";
import TextareaInput from "../../TextareaInput";
import FloatInput from "../../FloatInput";

/**
 * @title Supports multiple Input Types
 * @description InlineEdit supports inputs other than TextInput, so long as the value of the input can be stingified.
 */
export default function InlineEditPencil() {
  const [number, setNumber] = useState(1234);
  const [text, setText] = useState(
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit."
  );

  return (
    <Group gap={20}>
      <div style={{width: 240}}>
        <Label value="Number Input" indicateRequired={true}>
          <InlineEdit value={number} onChange={setNumber}>
            {/* $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0) */}
            {props => <FloatInput {...props} />}
          </InlineEdit>
        </Label>
      </div>

      <div style={{width: 240}}>
        <Label value="Text area Input" indicateRequired={true}>
          <InlineEdit
            editStyle="pencil"
            textWrap={true}
            value={text}
            onChange={setText}
          >
            {/* $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0) */}
            {props => <TextareaInput {...props} rows={{min: 2, max: 10}} />}
          </InlineEdit>
        </Label>
      </div>
    </Group>
  );
}
