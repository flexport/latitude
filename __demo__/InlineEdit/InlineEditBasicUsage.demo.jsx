/**
 * TEAM: frontend_infra
 * @flow
 */

import React, {useState} from "react";

import InlineEdit from "../../InlineEdit";
import Label from "../../Label";
import TextInput from "../../TextInput";

/**
 * @title Basic Usage
 * @description Have a value be editable directly in on the page using Inline Edit. When the `edit` button is pressed, the user will be prompted to edit the value.
 */
export default function InlineEditBasicUsage() {
  const [val, setVal] = useState("Theseus");

  return (
    <div style={{width: 240}}>
      <Label value="Ship Name" indicateRequired={true}>
        <InlineEdit value={val} onChange={setVal}>
          {/* $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0) */}
          {props => <TextInput {...props} />}
        </InlineEdit>
      </Label>
    </div>
  );
}
