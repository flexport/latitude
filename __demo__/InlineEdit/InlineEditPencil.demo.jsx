/**
 * TEAM: frontend_infra
 * @flow
 */

import React, {useState} from "react";

import InlineEdit from "../../InlineEdit";
import Label from "../../Label";
import TextInput from "../../TextInput";

/**
 * @title Pencil Inline Edit
 * @description InlineEdit also supports `pencil` style inline editting. In this mode, a pencil icon takes the place of the `Edit` Button.
 */
export default function InlineEditPencil() {
  const [val, setVal] = useState("FLEX001234");

  return (
    <div style={{width: 240}}>
      <Label value="Shipment ID" indicateRequired={true}>
        <InlineEdit
          editStyle="pencil"
          textWrap={true}
          value={val}
          onChange={setVal}
        >
          {/* $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0) */}
          {props => <TextInput {...props} />}
        </InlineEdit>
      </Label>
    </div>
  );
}
