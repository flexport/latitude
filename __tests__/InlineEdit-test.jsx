/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {mount} from "enzyme";
import {act} from "react-dom/test-utils";
import InlineEdit from "../InlineEdit";
import TextInput from "../TextInput";

jest.useFakeTimers();

describe("InlineEdit", () => {
  it("matches snapshot", () => {
    const comp = mount(
      <InlineEdit value="value" onChange={() => {}}>
        {/* $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0) */}
        {props => <TextInput {...props} />}
      </InlineEdit>
    );

    expect(comp).toMatchSnapshot();
  });

  describe("switching between edit and display states", () => {
    it("initially mounts display text", () => {
      const comp = mount(
        <InlineEdit value="value" onChange={() => {}}>
          {/* $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0) */}
          {props => <TextInput {...props} />}
        </InlineEdit>
      );

      expect(comp.find("TextInput").exists()).toBe(false);
      expect(comp.find("Text").exists()).toBe(true);
    });

    it("mounts input when the edit button is pressed", () => {
      const comp = mount(
        <InlineEdit value="value" onChange={() => {}}>
          {/* $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0) */}
          {props => <TextInput {...props} />}
        </InlineEdit>
      );

      comp.find("Button").simulate("click");
      act(() => {
        jest.runAllTimers();
      });
      comp.update();

      expect(comp.find("TextInput").exists()).toBe(true);
    });
  });

  describe("edit to display switching interactions", () => {
    let comp;
    let handleChange;

    beforeEach(() => {
      handleChange = jest.fn();

      comp = mount(
        <InlineEdit value="value" onChange={handleChange}>
          {/* $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0) */}
          {props => <TextInput {...props} />}
        </InlineEdit>
      );

      comp.find("Button").simulate("click");
      act(() => {
        jest.runAllTimers();
      });
      comp.update();
    });

    it("does not call onChange on new value when cancel is pressed", () => {
      comp.simulate("change", {target: {value: "newValue"}});
      expect(handleChange).not.toBeCalled();

      const cancelButton = comp.find({iconName: "cancel", type: "button"});

      cancelButton.simulate("click");
      expect(handleChange).not.toBeCalled();
    });

    it("does call onChange on new value and save is pressed", () => {
      comp.simulate("change", {target: {value: "newValue"}});
      expect(handleChange).not.toBeCalled();

      const checkButton = comp.find({iconName: "check", type: "button"});

      checkButton.simulate("click");
      expect(handleChange).toBeCalled();
    });
  });
});
