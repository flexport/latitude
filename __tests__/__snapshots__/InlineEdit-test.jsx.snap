// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineEdit matches snapshot 1`] = `
<InlineEdit
  onChange={[Function]}
  value="value"
>
  <div
    className="container_mgoi7r"
    style={Object {}}
  >
    <div
      className="editButtonValueContainer_3hvt2y-o_O-noWrap_vjioml"
    >
      <Text
        color="grey60"
      >
        <p
          className="standardOverride_eiiwmj"
          style={
            Object {
              "color": "#39414D",
              "display": "inline-block",
              "fontFamily": Array [
                "gt-america",
                "-apple-system",
                "BlinkMacSystemFont",
                "Segoe UI",
                "Roboto",
                "Oxygen",
                "Ubuntu",
                "Cantarell",
                "Fira Sans",
                "Droid Sans",
                "Helvetica Neue",
                "sans-serif",
              ],
              "fontSize": "14px",
              "fontStyle": "normal",
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": "20px",
              "overflow": "visible",
              "textOverflow": "initial",
              "textTransform": "initial",
              "whiteSpace": "inherit",
              "wordBreak": "normal",
            }
          }
        >
          value
        </p>
      </Text>
    </div>
    <div
      className="editButtonContainer_nukbix"
    >
      <Button
        disabled={false}
        intent="basic"
        isLoading={false}
        kind="blank"
        onClick={[Function]}
        size="m"
        type="button"
        width="responsive"
      >
        <button
          className="base_i7qsly-o_O-m_12juzcr-o_O-blank_1kmb22z-o_O-responsive_zk4wup"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <div
            className="base_sj570c-o_O-blank_1y3wr5v"
          >
            Edit
          </div>
        </button>
      </Button>
    </div>
  </div>
</InlineEdit>
`;
